### PR TITLE
remove sort requirement from pad-sequence

### PR DESCRIPTION
**Pull request by alexholdenmiller**
pad-sequence can get the max_len from the list of sequences.
entries only need to be sorted if output will be used for pack_padded_sequence, which can throw the error itself.

this way pad_sequence does O(n) work for me instead of me having to do O(nlogn) work in order to call the function.

thoughts?

### CODE REVIEW

File: torch/nn/utils/rnn.py [hunk 1]
```diff
@@ -251,10 +251,6 @@ def pad_sequence(sequences, batch_first=False, padding_value=0):
     out_variable = Variable(sequences[0].data.new(*out_dims).fill_(padding_value))
     for i, variable in enumerate(sequences):
         length = variable.size(0)
-        # temporary sort check, can be removed when we handle sorting internally
-        if prev_l < length:
-            raise ValueError("lengths array has to be sorted in decreasing order")
```
Comment (apaszke) [inline torch/nn/utils/rnn.py]: Removing this isn't sufficient. E.g. `pad_packed_sequence` is written with this assumption in mind, and will give you incorrect results.

Comment (SsnL) [reply torch/nn/utils/rnn.py]: Not really. The `PackPadded` op does the check itself https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/packing.py#L36.

**Comment by SsnL** [comment]
I feel this is a reasonable change. There is no reason to assume that `pad_sequence` is only for passing into a pytorch official RNN.

File: torch/nn/utils/rnn.py (hunk 1 above)
Comment (apaszke) [reply torch/nn/utils/rnn.py]: Right, I thought that was the function that was edited, my bad. However, I still think this is incorrect. You can't use the "batched copy" we're doing just 3 lines below if the sequences aren't sorted. You'd need a more sophisticated scattering for this.

**Review by apaszke — COMMENTED** [review]
We'll also need a test that checks it's ok to run this function with unsorted sequences, and the docstring will need an update too.

File: torch/nn/utils/rnn.py (hunk 1 above)
Comment (zou3519) [reply torch/nn/utils/rnn.py]: @apaszke how does the "batched copy" 3 lines below break if the sequences aren't sorted? It doesn't look like it depends on the order of the tensor. The only constraint is that `length` should be less than or equal to `max_len`.

Comment (apaszke) [reply torch/nn/utils/rnn.py]: The `:length` part is an issue. We can do this because we know that the tensor will always contain data in a contiguous subset of this dim. If the sequences aren’t sorted, we’ll need to maintain a list of indices

Comment (zou3519) [reply torch/nn/utils/rnn.py]: @apaszke Sorry, I'm not following the explanation (I think I'm misunderstanding how `pad_sequence` works).

If I do something like
```
import torch
from torch.nn.utils.rnn import pad_sequence
a = torch.full((1, 2), 0)
b = torch.full((2, 2), 1)
c = torch.full((3, 2), 2)
# Not in order
out = pad_sequence([a, c, b])
```
I expect `out[:,0]` to be a padded version of `a`, `out[:,1]` to be a padded version of `c`, `out[:,2]` to be a padded version of `b`. i.e.,
```
In [10]: out[:,2]
Out[10]:
tensor([[ 2.,  2.],
        [ 2.,  2.],
        [ 0.,  0.]])
```

That is what `out_variable[:length, i, ...] = variable` should give. What am I missing here?

**Comment by soumith** [comment]
merged via https://github.com/pytorch/pytorch/pull/7928
